src/librustc/ty/mod.rs

```diff
@@ -167,6 +167,16 @@ pub struct ImplHeader<'tcx> {
     pub predicates: Vec<Predicate<'tcx>>,
 }
 
+#[derive(Copy, Clone, PartialEq, RustcEncodable, RustcDecodable, HashStable)]
+pub enum ImplPolarity {
+    /// `impl Trait for Type`
+    Positive,
+    /// `impl !Trait for Type`
+    Negative,
+    /// `#[rustc_reservation_impl] impl Trait for Type`
+    Reservation,
```

Comment:
One last nit: Can you cite the tracking issue https://github.com/rust-lang/rust/issues/64631 here?